### PR TITLE
GH-2197: GraphTxn.forEachRemaining and .close

### LINKED ISSUE #2197: GraphTxn.find().forEachRemaining does not close the iterator.
### Version

5.0.0 development

### What happened?

GraphTxn.find returns an etxended iterator that does not get closed properly if exhaused by `Iterator.forEachRemaining`.

`WrappedGraph.forEachRemaining` call the wrapped class directly.

### Relevant output and stacktrace

```shell
public static void main(String... args) {
        Graph g = new GraphTxn();
        ExtendedIterator<Triple> iter = g.find();
        iter.toList();
        //iter.close();
        g.add(SSE.parseTriple("(:s :p :o)"));
    }
```

### Are you interested in making a pull request?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #2197

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx, or if in JIRA, JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

### CODE REVIEW

**Review by rvesse — APPROVED** [review]
LGTM

**Review by TelicentPaul — APPROVED** [review]

**Comment by afs** [comment]
Actually, draft did indeed mean draft in this case :-) 

Some tests need to be added. The changes to GraphTxn should be complete ... subject to the tests passing.

File: jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        graph.getTransactionHandler().execute(()->graph.add(triple1));
+        assertEquals(1,  graph.size());
+        ExtendedIterator<Triple> eIter = graph.find();
+        assertTrue(eIter.hasNext());
+        eIter.next();
+        // After .next, Not yet closed
+        eIter.hasNext();
+        // Now closed.
+        // The implementation, DatasetGraphInMemory doesn't require transactions.
+        graph.add(triple2);
```
Comment (kinow) [inline jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:54]: Would it be useful to test that triple2 was added or that graph.size is now 2 (is it committed/transaction closed here? Maybe 1?)? Or is it testing just that no error is raised?

File: jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        } finally { eIter.close(); }
+        // Now closed.
+        // The implementation, DatasetGraphInMemory doesn't require transactions.
+        graph.add(triple2);
+    }
+
+    @Test public void graphSteram01() {
+        GraphTxn graph = GraphFactory.createTxnGraph();
+        graph.add(triple2);
+        graph.stream().count();
```
Comment (kinow) [inline jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:75]: The count here is just to test that no error is raised?

**Review by kinow — APPROVED** [review]
Code looks good. I don't know the API well enough to say whether tests look good, so left a few comments (sorry if silly questions :+1:).

File: jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java (hunk 1 above)
Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:54]: These aren't full test of the graph. `GraphTxn` is wrapper around `DatasetGraphInMemory.getDefaultGraph` so the general "graphness" is that of the wrapped graph.

The tests are for the added features - the transactions, both old style jena-core and current transactionals, how they interact, and the support for `ExtendedIterator` as came up in #2197.

File: jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java (hunk 2 above)
Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:75]: `count` consumes and ends a stream. Any consuming terminal will do.

It causes the stream to be completely read. Behind the scenes, a stream from a GraphTxn is provided by an iterator and that which should call `GraphTxn.IteratorTxn.close`. `close` ends the READ transaction. `graph.add` fails if the READ transaction of iteration `graph.find` is still active.

Comment (kinow) [reply jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:75]: Thanks for explaining it!!

Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/sparql/graph/TestGraphTxn.java:75]: No problem!

**Pull request merged**